### PR TITLE
Add full-transaction test vectors

### LINKED ISSUE #24: Rationale for excluding cloneable
#### NFT Type

Categorized outputs may encode one of 5 NFT states:

- `NFT_NULL = 0x00`, *(b00000000)*
- `NFT_TRANSFER = 0x3F`, *(b00111111)* ("no capability", "immutable")
- `NFT_EDIT = 0x7F`, *(b10111111)* ("mutable")
- `NFT_CLONE = 0xBF`, *(b01111111)*
- `NFT_MINT = 0xFF`. *(b11111111)*

All other values are reserved and reading a reserved value will immediately fail the transaction.

### LINKED ISSUE #25: Rationale: `OP_UTXOTOKENCOMMITMENT` and `OP_OUTPUTTOKENCOMMITMENT` push `0` for zero-length commitments
"If the UTXO includes a non-fungible token with a zero-byte commitment, push 0x00. If the UTXO does not include a non-fungible token, push a 0 (VM Number)."

If we encode a 1-byte commitment that encodes a `commitment[0]=0x00` it would result in the same 0x00 value being pushed to the stack.

Suggestion: push an empty stack item in both no-NFT and no-commitment cases, and the user should check the capability to distinguish no-NFT from no-commitment.

### LINKED ISSUE #27: Improve `Shared Codepoint for All Tokens` Rationale
The [`Shared Codepoint for All Tokens`](https://github.com/bitjson/cashtokens/tree/f8b500a051f82d42dbf9e9e890bc6cdc14592307#shared-codepoint-for-all-tokens) rationale was written for a pre-release iteration of the CHIP and has gotten progressively weaker. (Now that we're [reusing the token inspection operation codepoints](https://github.com/bitjson/cashtokens/issues/7#issuecomment-1065480447) to indicate a token prefix, additional encodings do not "cost" instruction set codepoints.)

 I'm starting to think that rather than the sort of compound encoding we're currently using (`PREFIX_TOKEN <category_id> <has_nft> [commitment_length commitment] <ft_amount>`), we should use multiple prefixes to avoid the `0x00` placeholders being used.

So we'd have encodings for each of:
1. Only fungible tokens – e.g. `PREFIX_FUNGIBLETOKENS <category_id> <ft_amount>`
2. Only a nonfungible token – e.g. `PREFIX_NONFUNGIBLETOKEN <category_id> <capability> <commitment_length> [commitment]`
3. Both – e.g. `PREFIX_TOKENS <category_id> <capability> <ft_amount> <commitment_length> [commitment]`

This saves 1 byte per output when encoding only fungible tokens or only a nonfungible token (likely far more common than instances where both types are held), and it makes the encodings themselves simpler: there's no "empty NFT" or "zero FTs" state, and the presence of FTs and NFTs can always be determined by only the prefix.

Previously I'd thought that a shared prefix would make adding support to indexers slightly simpler, but in reality, indexers will almost certainly "translate" token encodings for their database; the encoding of tokens in transactions should just focus on being as simple and efficient as possible.

### LINKED ISSUE #32: Rationale for introducing token-aware CashAddress type bits
Suggested rationale

### Wallet Address Type

Because we require transactions involving categorized outputs to be signed using modified signing serialization, funds unintentionally sent to non-upgraded wallets can't be accidentally lost and will be recoverable by importing the recipient's wallet into upgraded software.

The existing [CashAddress](https://github.com/bitcoincashorg/bitcoincash.org/blob/master/spec/cashaddr.md#payload) standard supports extending the address through the use of a type bit, which could be used to indicate wallet support for receiving tokens:

>Encoding the size of the hash in the version field ensure that it is possible to check that the length of the address is correct.
>
>| Type bits |      Meaning      | Version byte value |
>| --------: | :---------------: | -----------------: |
>|         0 |       P2KH        |                  0 |
>|         1 |       P2SH        |                  8 |
>
>Further types will be added as new features are added.

However we expect that, over time, most wallets will support receiving and sending tokens.
**Introducing two more address types just to signal readiness to receive tokens would only have a temporary benefit, while the cost of having to maintain two more address encodings would be permanent.**

Therefore, we do NOT recommend a new address type.

### Risks - Wrong Wallet Implementation

If a wallet would obtain a categorized output but be unable to spend it, then it could cause it to temporarily become unusable until the issue is resolved.

This could happen in two ways:

1. Non-upgraded and badly implemented wallet registering a categorized output as one of its own, which is unlikely and [has been discussed](#script-prefix-on-outputs);
2. Upgraded wallet having a bug in transaction building, which can be expected since bugs happen, and not all wallets pass through the same quality assurance processes.

Burning funds by non-upgraded wallets will not be possible since they wouldn't be able to produce a valid signature, and it would at worst result in temporary loss of service - users unable to send their funds until the issue is resolved.
The same would hold for upgraded wallets that are upgraded but have a bug that would prevent them from producing valid transactions.

The riskiest class of bugs is that of an upgraded wallet able to produce signatures but failing to properly account amounts to create categorized "change" outputs that would capture amounts not intended to be spent.
This risk was already present with pure BCH wallet implementations, where the result would be a donation to miners in form of fees.
With this proposal, the risk applies the same to token amounts - but instead of being donated they'd be burned.

This is a risk external to protocol development, and the only mitigation available is to write good documentation and suggest using wallet software that has good quality-assurance processes.
Since most popular wallets have proper processes in place and upgrades are reviewed and tested by multiple developers, we consider this risk to be low.

### LINKED ISSUE #47: TOKEN_PREFIX byte needs to be a unique, unused and invalid op-code; disallow creation of locking scripts with prefix that failed to parse
This suggestion directly contradicts issue #41 and PR #43.  

If this suggestion were implemented it would simplify and/or reduce the "urgency" and/or criticality of #44, #45 and #46.

Suggestion Summary
---

I think it’s not a great idea for prefix byte to do double duty with the introspection opcode `OP_UTXOTOKENAMOUNT`.  We need to make the `TOKEN_PREFIX` be a unique codepoint that is otherwise an invalid opcode.  See discussion starting with my comment here:  https://github.com/bitjson/cashtokens/issues/41#issuecomment-1193218237

Suggest: `0xbf` as the token prefix .

Additionally: Post-activation, we should **disallow** the creation of locking scripts that start with `0xbf` but fail to parse as token data -- to save potential headaches in the future should we somehow loosen the serialization rules or otherwise modify them for the token data.

Rationale
---

While implementing #41, I noticed a couple of corner cases arise if we have the PREFIX_BYTE do double-duty as an op-code.  

- It means that we need to be **very careful** about what is a **consensus failure** and what is a **serialization failure**.    
  - This is because if we fail to parse a byte stream as token data (i.e. a **serialization failure**), the byte stream just ends up entirely put into the beginning of a `scriptPubKey` (locking script) of an output.  And entirely different consensus rules now apply to that data, as opposed to an output containing token data.  Thus, it puts a lot of burden on this specification to clearly delineate what is a **parse error** versus what is disallowed for **consensus**.  See issue #44 
- What's more, some locking scripts may unintentionally get "parsed" as token data when they intended to just have `0xd0` be their first op-code (very rare corner case, requires miners to mine these, but the fact that this is possible is troubling).... This "quirk" thus makes some subset of the set of all possible theoretical locking scripts beginning with `0xd0` impossible to construct.

If we use a **unique codepoint** that does **no double-duty** as an op-code (and is a disabled code-point otherwise),.. the above caveats *completely evaporate*.

---

Update: This has been implemented already here: https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1580/diffs?commit_id=874f825c8f33b19327cf6c55b62782b62548215f

### LINKED ISSUE #57: Specify that `SIGHASH_UTXOS` must also have `SIGHASH_FORKID` otherwise sigHashType is invalid
While this is implicit because of the fact that `SIGHASH_FORKID` is always required after August 2017, it would be nice to also see this in the specification as an explicitly stated thing.

This would avoid some implementations from e.g. forgetting to check that this flag must be active and somehow accepting "old" blocks that were invalid but that happened to use `0x20` -- this concern I have is imaginary really because we checkpoint in our nodes but -- just in case, it is worth mentioning in the spec.

### PR DESCRIPTION
**Pull request by bitjson**
Fixes #24
Fixes #25
Fixes #27 
Fixes #32
Fixes #47
Fixes #57

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -825,6 +830,62 @@ CashTokens differs from SLP in that CashTokens are validated by the network, and
...
+
+To use these test vectors, implementations should decode the transaction under test (5) and its UTXOs (6), then validate the transaction using all of the implementation's transaction validation infrastructure initialized in the expected standard/nonstandard mode(s). See [Implementations](#implementations) for examples.
+
+## Implementations
+
+Please see the following implementations for additional examples and test vectors:
+
+- C++
+  - [Bitcoin Cash Node (BCHN)](https://bitcoincashnode.org/) – A professional, miner-friendly node that solves practical problems for Bitcoin Cash.
+    - CashTokens support: [Merge Request !1580](https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1580)
```
Comment (cculianu) [inline readme.md:883]: Might want to link to the other two relevant ones that these tests depend on to be fully validatable:

- the p2psh32 one -- https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1556
- the tx-size one (maybe?) -- https://gitlab.com/bitcoin-cash-node/bitcoin-cash-node/-/merge_requests/1598

**Comment by cculianu** [comment]
Two of the tests: `"yv08c"` and `"n599r"` contain bad input coins -- dupe prevout:N in the tx.vin.  Please fix this as it fails BCHN validation when a tx is "broken" like this.

**Comment by cculianu** [comment]
More issues  with the **standard** tests:

- `"n4g8m"` - spends and/or creates dust (< 576 satoshis) -- this is NON standard.
- `"l0s94"` - this tx is malformed/out-of-consensus because inputs are below outputs (in BCH)
- `"x7ryc"` - this tx fails because it contains a genesis that is creating a "new" token but the input coin already has the same identical token on it (same category id as the prevout tx id!!).  This is impossible in real code but BCHN checks for this condition because it would be a bad bug.  This is malformed in BCHN.  More specifically: 
```
Input 0: 0000000000000000000000000000000000000000000000000000000000000001:0, tokenData = token::OutputData(id=0000000000000000000000000000000000000000000000000000000000000001, bitfield=10, amount=1000, commitment=)
Output 0: CTxOut(nValue=0.00001000, scriptPubKey=a914da1745e9b549bd0bfa1a569971 token::OutputData(id=0000000000000000000000000000000000000000000000000000000000000001, bitfield=10, amount=1, commitment=))
```

**Comment by cculianu** [comment]
Non-standard test: `"9r8vf"` has the same "in-below-out" problem.

**Comment by cculianu** [comment]
Similarly, non-standard test txn `"72fht"` has the same "dupe genesis" problem as `"x7ryc"` in the standard tests.

**Comment by cculianu** [comment]
Non-standard test txn `"j7wwg"` has the same "duplicated input" problem as ["yv08c" and "n599r"](https://github.com/bitjson/cashtokens/pull/58#issuecomment-1221525304).

**Comment by cculianu** [comment]
"Invalid, pre-activation test" has the following defects:

- various txns that have duplicated inputs, in addition to the ones above with dupe inputs (which also exist in this test), we have: `"lg530"`, `"vmvk9"`, `"gzmc5"`
- PATFO-creating txns should be are accepted as nonstandard pre-activation. They should all be allowed (even though they are unspendable forever pre-and-post activation)... but your test vector has some and they are wrongly prescribed to be rejected.
  - `"4g0e5"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)
  - `"q4dd7"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)
  - `"ps0z3"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)
  - `"7mnux"` is a PATFO (should be accepted as nonstandard, but test vector assumes wrongly that it should be rejected)

**Comment by cculianu** [comment]
Some txns like `"slpwl"` create too-small a BCH output amount for the tx size (note that DUST is not always 546 -- it is a complicated equation based off txn size).  So they get flagged as violating "dust limist" and rejected in standard mode, but are accepted in non-standard .. for instance..

**Comment by bitjson** [comment]
Thanks again for finding these issues @cculianu! I'll try to have these fixed soon. (Please continue to make notes here if you notice other issues!)

**Comment by cculianu** [comment]
Awesome, thanks.  Note there may be other issues with the p2sh32 txns -- aside from unserializing those I haven't run them through the tests because the CashTokens MR and the p2sh32 MR in BCHN are separate works and not merged together yet.

I suspect it's possible the p2sh32 tests also suffer from at least some of the same pattern of errors above (in-below-out, creates dust, duped genesis, etc).  So while you go back and fix perhaps retrace your steps and see if those p2sh32 txns also may be susceptible to the same pattern of error..?

**Comment by cculianu** [comment]
Also note that some of the txns are undersized (<65 bytes) but are in test vectors that are expected to pass consensus, in particular:

- `dt9lw`, `xsv0y`, `auqfw`, `e393g`, `ezf3r`

There may be others (in the p2sh32 tests which I don't even validate yet).

**Comment by cculianu** [comment]
Some p2sh32 txns have errors

- `d6dqw` (dupe genesis id)
- `ks078` (dust)

**Comment by bitjson** [comment]
Thanks @cculianu and @Andrew-128 for reviewing test vectors against the BCHN implementation! 

@cculianu – I think I've corrected all of the issues above. Please let me know if you notice others.
@Andrew-128 – thanks for the [PR](https://github.com/bitauth/libauth/pull/99)! I pulled those new test vectors into this repo too.

I also just added a few sets of test vectors covering signing serializations and dust threshold calculation (since it's sensitive to token encoding). I'd still like to get a bit more coverage over signing serializations, but I think this PR is about ready to merge. (And I'll send those additional test vectors in another PR.)

Is everyone ok with this being merged now?

**Comment by cculianu** [comment]
@bitjson Thanks for this. I'll integrate these tomorrow into the tests we already have and see if they work or if they blow up. I'll get back to you tomorrow on this.

**Comment by bitjson** [comment]
Ok, I just added the rest of the signing serialization tests I'd been planning: signing with every available algorithm over several more transaction constructions and every type of token encoding + some malleability tests. 

So with that, I think the spec is thoroughly covered! Ready to merge once we've confirmed these pass on BCHN's implementation. 👍

**Comment by bitjson** [comment]
Thanks @Andrew-128 for confirming these tests are all working on BCHN! 🚀

I'll go ahead and merge to get these changes in master, and we can make any needed corrections in smaller PRs.

**Pull request merged**